README.md
# Introduction

[![travis](https://img.shields.io/travis/adrai/node-cqrs-saga.svg)](https://travis-ci.org/adrai/node-cqrs-saga) [![npm](https://img.shields.io/npm/v/cqrs-saga.svg)](https://npmjs.org/package/cqrs-saga)

Node-cqrs-saga is a node.js module that helps to implement the sagas in cqrs.
It can be very useful as domain component if you work with (d)ddd, cqrs, eventdenormalizer, host, etc.

# Installation

    $ npm install cqrs-saga


# Usage

	var pm = require('cqrs-saga')({
	  // the path to the "working directory"
	  // can be structured like
	  // [set 1](https://github.com/adrai/node-cqrs-saga/tree/master/test/integration/fixture)
	  sagaPath: '/path/to/my/files',

	  // optional, default is 800
	  // if using in scaled systems and not guaranteeing that each event for a saga "instance"
	  // dispatches to the same worker process, this module tries to catch the concurrency issues and
	  // retries to handle the event after a timeout between 0 and the defined value
	  retryOnConcurrencyTimeout: 1000,

	  // optional, default is in-memory
	  // currently supports: mongodb, redis, azuretable and inmemory
	  // hint settings like: [eventstore](https://github.com/adrai/node-eventstore#provide-implementation-for-storage)
	  // mongodb:
	  sagaStore: {
	    type: 'mongodb',
	    host: 'localhost',                          // optional
	    port: 27017,                                // optional
	    dbName: 'domain',                           // optional
	    collectionName: 'sagas',                    // optional
	    timeout: 10000                              // optional
      // authSource: 'authedicationDatabase',        // optional
	    // username: 'technicalDbUser',                // optional
	    // password: 'secret'                          // optional
      // url: 'mongodb://user:pass@host:port/db?opts // optional
	  },
	  // or redis:
	  sagaStore: {
	    type: 'redis',
	    host: 'localhost',                          // optional
	    port: 6379,                                 // optional
	    db: 0,                                      // optional
	    prefix: 'domain_saga',                      // optional
	    timeout: 10000                              // optional
	    // password: 'secret'                          // optional
	  },

	  // optional, default is in-memory
	  // the revisionguard only works if aggregateId and revision are defined in event definition
	  // currently supports: mongodb, redis, tingodb, azuretable and inmemory
	  // hint settings like: [eventstore](https://github.com/adrai/node-eventstore#provide-implementation-for-storage)
	  revisionGuard: {
	  	queueTimeout: 1000,                         // optional, timeout for non-handled events in the internal in-memory queue
	  	queueTimeoutMaxLoops: 3,                    // optional, maximal loop count for non-handled event in the internal in-memory queue
      startRevisionNumber: 1,			                // optional, if defined the denormaizer waits for an event with that revision to be used as first event

	  	type: 'redis',
	  	host: 'localhost',                          // optional
	  	port: 6379,                                 // optional
	  	db: 0,                                      // optional
	  	prefix: 'saga_revision',               			// optional
	  	timeout: 10000                              // optional
	  	// password: 'secret'                       // optional
	  }
		// or, if you don't like to guard revisions ( managed externally ? )
		revisionGuard: false // note has to be set explicitly false
	});


## Catch connect ad disconnect events

	// sagaStore
	pm.sagaStore.on('connect', function() {
	  console.log('sagaStore connected');
	});

	pm.sagaStore.on('disconnect', function() {
	  console.log('sagaStore disconnected');
	});

	// revisionGuardStore
	pm.revisionGuardStore.on('connect', function() {
	  console.log('revisionGuardStore connected');
	});

	pm.revisionGuardStore.on('disconnect', function() {
	  console.log('revisionGuardStore disconnected');
	});


	// anything (sagaStore or revisionGuardStore)
	pm.on('connect', function() {
	  console.log('something connected');
	});

	pm.on('disconnect', function() {
	  console.log('something disconnected');
	});


## Define the event structure
The values describes the path to that property in the event message.

	pm.defineEvent({
	  // optional, default is 'name'
	  name: 'name',

	  // optional, only makes sense if contexts are defined in the 'domainPath' structure
	  context: 'context.name',

	  // optional, only makes sense if aggregates with names are defined in the 'domainPath' structure
	  aggregate: 'aggregate.name',

	  // optional, default is 'aggregate.id'
	  aggregateId: 'aggregate.id',

	  // optional, default is 'revision'
	  // will represent the aggregate revision, can be used in next command
	  revision: 'revision',

	  // optional
	  version: 'version',

	  // optional, if defined theses values will be copied to the command (can be used to transport information like userId, etc..)
	  meta: 'meta'
	});


## Define the command structure
The values describes the path to that property in the command message.

	pm.defineCommand({
	  // optional, default is 'id'
	  id: 'id',

	  // optional, if defined the values of the event will be copied to the command (can be used to transport information like userId, etc..)
	  meta: 'meta'
	});


## Define the id generator function [optional]
### you can define a synchronous function

	pm.idGenerator(function () {
	  var id = require('uuid').v4().toString();
	  return id;
	});

### or you can define an asynchronous function

	pm.idGenerator(function (callback) {
	  setTimeout(function () {
	    var id = require('uuid').v4().toString();
	    callback(null, id);
	  }, 50);
	});


## Wire up commands [optional]
### you can define a synchronous function

	// pass commands to bus
	pm.onCommand(function (cmd) {
	  bus.emit('command', cmd);
	});

### or you can define an asynchronous function

	// pass commands to bus
	pm.onCommand(function (cmd, callback) {
	  bus.emit('command', cmd, function ack () {
	    callback();
	  });
	});


## Wire up event missing [optional]

	pm.onEventMissing(function (info, evt) {

		// grab the missing events, depending from info values...
		// info.aggregateId
		// info.aggregateRevision
		// info.aggregate
		// info.context
		// info.guardRevision
		// and call handle...
		pm.handle(missingEvent, function (err) {
			if (err) { console.log(err); }
		});

	});

you can get the last guarded event:

	pm.getLastEvent(function (err, evt) {

	  if (event.occurredAt < Date.now()) {
	  	// ...
	  }

	});

## Using custom structure loader function
The built-in structure loader can be replaced with one adapted to your needs.
To do that, you need to include a loading method in the options object passed to the domain constructor.

	// options will contain sagaPath as well as the as well as a definition object containing all the constructors of the saga components  ( Saga )
	function structureLoader(options) {

		return [
			new options.definitions.Saga({
				name: 'myEvt'
			}, function (evt, saga, callback) {
				// handle
			}),
			new options.definitions.Saga({
				name: 'myOtherEvt'
			}, function (evt, saga, callback) {
				// handle
			}),
		];
		// or more probably
		return myExternalLoader(options.sagaPath, options.definitions);
	}

	var pm = require('cqrs-saga')({
	  sagaPath: '/path/to/my/files',
		structureLoader: structureLoader
	});

## Initialization

	pm.init(function (err, warnings) {
	  // this callback is called when all is ready...
	  // warnings: if no warnings warnings is null, else it's an array containing errors during require of files
	});

	// or

	pm.init(); // callback is optional


## Handling an event

	pm.handle({
	  id: 'b80ade36-dd05-4340-8a8b-846eea6e286f',
	  name: 'orderCreated',
	  aggregate: {
	    id: '3b4d44b0-34fb-4ceb-b212-68fe7a7c2f70',
	    name: 'order'
	  },
	  context: {
	    name: 'sale'
	  },
	  payload: {
      totalCosts: 520,
      seats: ['4f', '8a']
	  },
	  revision: 0,
	  version: 1,
	  meta: {
	    userId: 'ccd65819-4da4-4df9-9f24-5b10bf89ef89'
	  }
	}); // callback is optional

### or

	pm.handle({
	  id: 'b80ade36-dd05-4340-8a8b-846eea6e286f',
	  name: 'orderCreated',
    aggregate: {
      id: '3b4d44b0-34fb-4ceb-b212-68fe7a7c2f70',
      name: 'order'
    },
    context: {
      name: 'sale'
    },
    payload: {
      totalCosts: 520,
      seats: ['4f', '8a']
    },
	  revision: 0,
	  version: 1,
	  meta: {
	    userId: 'ccd65819-4da4-4df9-9f24-5b10bf89ef89'
	  }
	}, function (errs, cmds) {
	  // this callback is called when event is handled successfully or unsuccessfully
	  // errs can be of type:
	  // - null
	  // - Array of Errors
	  //
	  // cmds: same as passed in 'onCommand' function
	});

### more infos, can be useful if testing

	pm.handle({
	  id: 'b80ade36-dd05-4340-8a8b-846eea6e286f',
	  name: 'orderCreated',
    aggregate: {
      id: '3b4d44b0-34fb-4ceb-b212-68fe7a7c2f70',
      name: 'order'
    },
    context: {
      name: 'sale'
    },
    payload: {
      totalCosts: 520,
      seats: ['4f', '8a']
    },
	  meta: {
	    userId: 'ccd65819-4da4-4df9-9f24-5b10bf89ef89'
	  }
	}, function (errs, cmds, sagaModels) {
	  // this callback is called when event is handled successfully or unsuccessfully
	  // errs: is the same as described before

	  // cmds: same as passed in 'onCommand' function
	  // cmds: in case of no error or in case of error here is the array of all commands that should be published

	  // sagaModels: represents the saga data after have handled the event
	});


## Request saga information

After the initialization you can request the saga information:

	pm.init(function (err) {
	  pm.getInfo();
	  // ==>
	  // {
	  //   "sagas": [
	  //     {
	  //       "name": "orderConfirmed",
	  //       "aggregate": "order",
	  //       "context": "sale",
	  //       "version": 0
	  //     },
	  //     {
	  //       "name": "orderCreated",
	  //       "aggregate": "order",
	  //       "context": "sale",
	  //       "version": 0
	  //     },
	  //     {
	  //       "name": "paymentAccepted",
	  //       "aggregate": "payment",
	  //       "context": "sale",
	  //       "version": 2
	  //     },
	  //     {
	  //       "name": "seatsReserved",
	  //       "aggregate": "reservation",
	  //       "context": "sale",
	  //       "version": 0
	  //     }
	  //   ]
	  // }
	});


# Components definition

## Saga

	module.exports = require('cqrs-saga').defineSaga({
	  // optional, default is file name without extension

releasenotes.md
## [v1.11.14](https://github.com/adrai/node-cqrs-saga/compare/v1.11.13...v1.11.14)
- mongodb: useUnifiedTopology

## [v1.11.13](https://github.com/adrai/node-cqrs-saga/compare/v1.11.12...v1.11.13)
- add option not to use the revisionGuardStore

## [v1.11.12](https://github.com/adrai/node-cqrs-saga/compare/v1.11.11...v1.11.12)
- replace additional deprecated mongo methods

## [v1.11.11](https://github.com/adrai/node-cqrs-saga/compare/v1.11.1...v1.11.11)
- remove debug forgotten console messages

## [v1.11.1](https://github.com/adrai/node-cqrs-saga/compare/v1.11.0...v1.11.1)
- replace deprecated mongo methods ( ensureIndex, insert, remove, update )
- add optional shouldHandleEvent function, to filter events before loading saga

## [v1.11.0](https://github.com/adrai/node-cqrs-saga/compare/v1.10.2...v1.11.0)
- add option to add custom structureLoader implementation

## [v1.10.2](https://github.com/adrai/node-cqrs-saga/compare/v1.10.1...v1.10.2)
- remove deprecated option max_attempts from redis options

## [v1.10.1](https://github.com/adrai/node-cqrs-saga/compare/v1.9.1...v1.10.1)
- compatibility with new mongodb version (3.x)

## [v1.9.1](https://github.com/adrai/node-cqrs-saga/compare/v1.9.0...v1.9.1)
- add possibility to call removeTimeout after getTimeoutSagas and getOlderSagas [#42](https://github.com/adrai/node-cqrs-saga/issues/#42) thanks to [edro](https://github.com/edro)

## [v1.9.0](https://github.com/adrai/node-cqrs-saga/compare/v1.8.0...v1.9.0)
- optional pagination for getTimeoutedSagas and getUndispatchedCommands

## [v1.8.0](https://github.com/adrai/node-cqrs-saga/compare/v1.7.3...v1.8.0)
- Support default exports [#40](https://github.com/adrai/node-cqrs-saga/pull/#40) thanks to [IRT-fbachmann](https://github.com/IRT-fbachmann)

## [v1.7.3](https://github.com/adrai/node-cqrs-saga/compare/v1.7.2...v1.7.3)
- fix saga callback on retry [#39](https://github.com/adrai/node-cqrs-saga/pull/#39) thanks to [nanov](https://github.com/nanov)

## [v1.7.2](https://github.com/adrai/node-cqrs-saga/compare/v1.7.1...v1.7.2)
- optimize handling for commandRejected

## [v1.7.1](https://github.com/adrai/node-cqrs-saga/compare/v1.6.27...v1.7.1)
- for getTimeoutedSagas handling: if the pm.onCommand handler is registered it will be automatically executed

## [v1.6.27](https://github.com/adrai/node-cqrs-saga/compare/v1.6.25...v1.6.27)
- fix for new mongodb driver

## [v1.6.25](https://github.com/adrai/node-cqrs-saga/compare/v1.6.24...v1.6.25)
- update deps

## [v1.6.24](https://github.com/adrai/node-cqrs-saga/compare/v1.6.23...v1.6.24)
- edgecase in revisionGuard

## [v1.6.23](https://github.com/adrai/node-cqrs-saga/compare/v1.6.22...v1.6.23)
- redis, mongodb: call disconnect on ping error

## [v1.6.22](https://github.com/adrai/node-cqrs-saga/compare/v1.6.21...v1.6.22)
- Fix events getting lost at high concurrency [#33](https://github.com/adrai/node-cqrs-saga/pull/#33) thanks to [hilkeheremans](https://github.com/hilkeheremans)

## [v1.6.21](https://github.com/adrai/node-cqrs-saga/compare/v1.6.20...v1.6.21)
- Support mongo connection string

## [v1.6.20](https://github.com/adrai/node-cqrs-saga/compare/v1.6.19...v1.6.20)
- redis, mongodb: call disconnect on ping error

## [v1.6.19](https://github.com/adrai/node-cqrs-saga/compare/v1.6.17...v1.6.19)
- revisionGuard: optional startRevisionNumber

## [v1.6.17](https://github.com/adrai/node-cqrs-saga/compare/v1.6.16...v1.6.17)
- redis: added optional heartbeat

## [v1.6.16](https://github.com/adrai/node-cqrs-saga/compare/v1.6.15...v1.6.16)
- fix defineShouldHandle

## [v1.6.15](https://github.com/adrai/node-cqrs-saga/compare/v1.6.14...v1.6.15)
- revisionGuard fix

## [v1.6.14](https://github.com/adrai/node-cqrs-saga/compare/v1.6.13...v1.6.14)
- redis: fix for new redis lib

## [v1.6.13](https://github.com/adrai/node-cqrs-saga/compare/v1.6.12...v1.6.13)
- mongodb: added optional heartbeat

## [v1.6.12](https://github.com/adrai/node-cqrs-saga/compare/v1.6.11...v1.6.12)
- redis: fix wrong multi response handling

## [v1.6.11](https://github.com/adrai/node-cqrs-saga/compare/v1.6.10...v1.6.11)
- optimize handling of guarding the first events

## [v1.6.10](https://github.com/adrai/node-cqrs-saga/compare/v1.6.9...v1.6.10)
- remove trycatch dependency due to memory leaks

## [v1.6.9](https://github.com/adrai/node-cqrs-saga/compare/v1.6.8...v1.6.9)
- give possibility to use mongodb with authSource

## [v1.6.8](https://github.com/adrai/node-cqrs-saga/compare/v1.6.7...v1.6.8)
- updated dep

## [v1.6.7](https://github.com/adrai/node-cqrs-saga/compare/v1.6.6...v1.6.7)
- optimization for `npm link`'ed development

## [v1.6.6](https://github.com/adrai/node-cqrs-saga/compare/v1.6.4...v1.6.6)
- catch throwing errors when calling callback

## [v1.6.4](https://github.com/adrai/node-cqrs-saga/compare/v1.6.3...v1.6.4)
- expose warnings during initialization

## [v1.6.3](https://github.com/adrai/node-cqrs-saga/compare/v1.6.2...v1.6.3)
- better catch for userland errors

## [v1.6.2](https://github.com/adrai/node-cqrs-saga/compare/v1.6.1...v1.6.2)
- fix alreadyInQueue check

## [v1.6.1](https://github.com/adrai/node-cqrs-saga/compare/v1.6.0...v1.6.1)
- redis: replace .keys() calls with .scan() calls => scales better


lib/pm.js
var debug = require('debug')('saga'),
  async = require('async'),
  util = require('util'),
  EventEmitter = require('events').EventEmitter,
  _ = require('lodash'),
  EventDispatcher = require('./eventDispatcher'),
  sagastore = require('./store'),
  uuid = require('uuid').v4,
  dotty = require('dotty'),
  RevisionGuard = require('./revisionGuard'),
  revisionGuardStore = require('./revisionGuardStore'),
  SagaModel = require('./sagaModel'),
  customLoader = require('./structure/customLoader'),
  structureLoader = require('./structure/structureLoader'),
  attachLookupFunctions = require('./structure/structureSearcher');


function createStructureLoader (options) {
  if (options) {
    if (!_.isFunction(options)) {
      var err = new Error('"options.structureLoader" is not a valid structure loader method');
      debug(err);
      throw err;
    };
    return customLoader(options);
  }
  return structureLoader;
}
  
/**
 * ProcessManagement constructor
 * @param {Object} options The options.
 * @constructor
 */
function ProcessManagement(options) {
  EventEmitter.call(this);

  options = options || {};

  if (!options.sagaPath && options.sagaPath !== '') {
    var err = new Error('Please provide sagaPath in options');
    debug(err);
    throw err;
  }

  var defaults = {
    commandRejectedEventName: 'commandRejected'
  };

  _.defaults(options, defaults);

  this.structureLoader = createStructureLoader(options.structureLoader);
  this.sagaStore = sagastore.create(options.sagaStore);

  options.retryOnConcurrencyTimeout = options.retryOnConcurrencyTimeout || 800;

  var defaultRevOpt = {
    queueTimeout: 1000,
    queueTimeoutMaxLoops: 3//,
    // startRevisionNumber: 1
  };

  if (options.revisionGuard !== false) {
    options.revisionGuard = options.revisionGuard || {};
    _.defaults(options.revisionGuard, defaultRevOpt);
    this.revisionGuardStore = revisionGuardStore.create(options.revisionGuard);
  }

  this.options = options;

  this.definitions = {
    command: {
      id: 'id'
////      name: 'name',                   // optional
////      aggregateId: 'aggregate.id',    // optional
////      context: 'context.name',        // optional
////      aggregate: 'aggregate.name',    // optional
////      payload: 'payload',             // optional
////      revision: 'revision',           // optional
////      version: 'version',             // optional
//      meta: 'meta'                    // optional (will be passed directly to corresponding event(s))
    },
    event: {
////      correlationId: 'correlationId', // optional
////      id: 'id',                       //optional
      name: 'name'
////      aggregateId: 'aggregate.id',    // optional
//      context: 'context.name',        // optional
//      aggregate: 'aggregate.name',    // optional
////      payload: 'payload',             // optional
////      revision: 'revision'            // optional
//      version: 'version',             // optional
//      meta: 'meta'                    // optional (will be passed directly from corresponding command)
    }
  };

  this.idGenerator(function () {
    return uuid().toString();
  });

  // DO not register a "dummy" onCommand handler because of getTimeoutedSagas handling
  // this.onCommand(function (cmd) {
  //   debug('emit:', cmd);
  // });

  this.onEventMissing(function (info, evt) {
    debug('missing events: ', info, evt);
  });
}

util.inherits(ProcessManagement, EventEmitter);

_.extend(ProcessManagement.prototype, {

  /**
   * Inject definition for command structure.
   * @param   {Object} definition the definition to be injected
   * @returns {ProcessManagement} to be able to chain...
   */
  defineCommand: function (definition) {
    if (!definition || !_.isObject(definition)) {
      var err = new Error('Please pass a valid definition!');
      debug(err);
      throw err;
    }

    this.definitions.command = _.defaults(definition, this.definitions.command);
    return this;
  },

  /**
   * Inject definition for event structure.
   * @param   {Object} definition the definition to be injected
   * @returns {ProcessManagement} to be able to chain...
   */
  defineEvent: function (definition) {
    if (!definition || !_.isObject(definition)) {
      var err = new Error('Please pass a valid definition!');
      debug(err);
      throw err;
    }

    this.definitions.event = _.defaults(definition, this.definitions.event);
    return this;
  },

  /**
   * Inject idGenerator function.
   * @param   {Function}  fn      The function to be injected.
   * @returns {ProcessManagement} to be able to chain...
   */
  idGenerator: function (fn) {
    if (!fn || !_.isFunction(fn)) {
      var err = new Error('Please pass a valid function!');
      debug(err);
      throw err;
    }

    if (fn.length === 1) {
      this.getNewId = fn;
      return this;
    }

    this.getNewId = function (callback) {
      callback(null, fn());
    };

    return this;
  },

  /**
   * Inject function for for event notification.
   * @param   {Function} fn       the function to be injected
   * @returns {ProcessManagement} to be able to chain...
   */
  onCommand: function (fn) {
    if (!fn || !_.isFunction(fn)) {
      var err = new Error('Please pass a valid function!');
      debug(err);
      throw err;
    }

    if (fn.length === 1) {
      fn = _.wrap(fn, function(func, cmd, callback) {
        func(cmd);
        callback(null);
      });
    }

    this.onCommandHandle = fn;

    return this;
  },

  /**
   * Inject function for event missing handle.
   * @param   {Function} fn       the function to be injected
   * @returns {ProcessManager} to be able to chain...
   */
  onEventMissing: function (fn) {
    if (!fn || !_.isFunction(fn)) {
      var err = new Error('Please pass a valid function!');
      debug(err);
      throw err;
    }

    this.onEventMissingHandle = fn;

    return this;
  },

  /**
   * Call this function to initialize the saga.
   * @param {Function} callback the function that will be called when this action has finished [optional]
   *                            `function(err){}`
   */
  init: function (callback) {

    var self = this;

    var warnings = null;

    async.series([
      // load saga files...
      function (callback) {
        if (self.options.sagaPath === '') {
          self.sagas = {};
          debug('empty sagaPath defined so no sagas will be loaded...');
          return callback(null);
        }
        debug('load saga files...');
        self.structureLoader(self.options.sagaPath, function (err, sagas, warns) {
          if (err) {
            return callback(err);
          }
          warnings = warns;
          self.sagas = attachLookupFunctions(sagas);
          callback(null);
        });
      },

      // prepare infrastructure...
      function (callback) {
        debug('prepare infrastructure...');
        async.parallel([

          // prepare sagaStore...
          function (callback) {
            debug('prepare sagaStore...');

            self.sagaStore.on('connect', function () {
              self.emit('connect');
            });

            self.sagaStore.on('disconnect', function () {
              self.emit('disconnect');
            });

            self.sagaStore.connect(callback);
          },

          // prepare revisionGuard...
          function (callback) {
            debug('prepare revisionGuard...');
            if (!self.revisionGuardStore) {
              return callback(null);
            }

            self.revisionGuardStore.on('connect', function () {
              self.emit('connect');
            });

            self.revisionGuardStore.on('disconnect', function () {
              self.emit('disconnect');
            });

            self.revisionGuardStore.connect(callback);
          }
        ], callback);
      },

      // inject all needed dependencies...
      function (callback) {
        debug('inject all needed dependencies...');

        if (self.revisionGuardStore) {
          self.revisionGuard = new RevisionGuard(self.revisionGuardStore, self.options.revisionGuard);
          self.revisionGuard.onEventMissing(function (info, evt) {
            self.onEventMissingHandle(info, evt);
          });
        }

        if (self.options.sagaPath !== '') {
          self.eventDispatcher = new EventDispatcher(self.sagas, self.definitions.event);
          self.sagas.defineOptions({}) // options???
            .defineCommand(self.definitions.command)
            .defineEvent(self.definitions.event)
            .idGenerator(self.getNewId)
            .useSagaStore(self.sagaStore);
        }

        if (self.revisionGuardStore) {
          self.revisionGuard.defineEvent(self.definitions.event);
        }

        callback(null);
      }
    ], function (err) {
      if (err) {
        debug(err);
      }
      if (callback) { callback(err, warnings); }
    });
  },

  /**
   * Returns the saga information.
   * @returns {Object}
   */
  getInfo: function () {
    if (!this.sagas) {
      var err = new Error('Not initialized!');
      debug(err);
      throw err;
    }

    return this.sagas.getInfo();
  },

  /**
   * Call this function to forward it to the dispatcher.
   * @param {Object}   evt      The event object
   * @param {Function} callback The function that will be called when this action has finished [optional]
   *                            `function(errs, evt, notifications){}` notifications is of type Array
   */
  dispatch: function (evt, callback) {
    var self = this;

    this.eventDispatcher.dispatch(evt, function (errs, sagaModels) {
      var cmds = [];

      if (!sagaModels || sagaModels.length === 0) {
        if (callback) {
          callback(errs, cmds, []);
        }
        return;
      }

      async.each(sagaModels, function (sagaModel, callback) {

        var cmdsToSend = sagaModel.getUndispatchedCommands();

        function setCommandToDispatched (c, clb) {
          debug('set command to dispatched');
          self.setCommandToDispatched(dotty.get(c, self.definitions.command.id), sagaModel.id, function (err) {
            if (err) {
              return clb(err);
            }
            cmds.push(c);
            sagaModel.removeUnsentCommand(c);
            clb(null);
          });
        }

        async.each(cmdsToSend, function (cmd, callback) {

          if (self.onCommandHandle) {
            debug('publish a command');
            self.onCommandHandle(cmd, function (err) {
              if (err) {
                debug(err);
                return callback(err);
              }
              setCommandToDispatched(cmd, callback);
            });
          } else {
            setCommandToDispatched(cmd, callback);
          }

        }, callback);

      }, function (err) {
        if (err) {
          if (!errs) {
            errs = [err];
          } else if (_.isArray(errs)) {
            errs.unshift(err);
          }
          debug(err);
        }
        if (callback) {
          try {
            callback(errs, cmds, sagaModels);
          } catch (e) {
            debug(e);
            console.log(e.stack);
            process.emit('uncaughtException', e);
          }
        }
      });

    });
  },

  /**
   * Call this function to let the saga handle it.
   * @param {Object}   evt      The event object
   * @param {Function} callback The function that will be called when this action has finished [optional]
   *                            `function(err, cmds, sagaModels){}` cmds and sagaModels are of type Array
   */
  handle: function (evt, callback) {
    if (!evt || !_.isObject(evt)) {
      var err = new Error('Please pass a valid event!');
      debug(err);
      throw err;
    }

    var self = this;

    var workWithRevisionGuard = false;
    if (
      this.revisionGuard &&
      !!this.definitions.event.revision && dotty.exists(evt, this.definitions.event.revision) &&
      !!this.definitions.event.aggregateId && dotty.exists(evt, this.definitions.event.aggregateId)
    ) {
      workWithRevisionGuard = true;
    }

    if (dotty.get(evt, this.definitions.event.name) === this.options.commandRejectedEventName) {
      workWithRevisionGuard = false;
    }

    if (!workWithRevisionGuard) {
      return this.dispatch(evt, callback);
    }

    this.revisionGuard.guard(evt, function (err, done) {
      if (err) {
        debug(err);
        if (callback) {
          callback([err]);
        }
        return;
      }

      self.dispatch(evt, function (errs, cmds, sagaModels) {
        if (errs) {
          debug(errs);
          if (callback) {
            callback(errs, cmds, sagaModels);
          }
          return;
        }

        done(function (err) {
          if (err) {
            if (!errs) {
              errs = [err];
            } else if (_.isArray(errs)) {
              errs.unshift(err);
            }
            debug(err);
          }

          if (callback) {
            callback(errs, cmds, sagaModels);
          }
        });
      });

    });

  },

  /**
   * Use this function to get all timeouted sagas.
   * @param  {Function} callback The function, that will be called when this action is completed.
   *                             `function(err, sagas){}` saga is of type Array.
   */
  getTimeoutedSagas: function (options, callback) {
    if (!callback) {
      callback = options;
      options = {};
    }

    var self = this;

    this.sagaStore.getTimeoutedSagas(options, function (err, sagas) {
      if (err) {
        debug(err);
        return callback(err);
      }

      var sagaModels = [];
      sagas.forEach(function (s) {
        var sagaModel = new SagaModel(s.id);
        sagaModel.set(s);
        sagaModel.actionOnCommit = 'update';

        var calledAddCommandToSend = false;
        sagaModel.addCommandToSend = function (cmd) {
          calledAddCommandToSend = true;
          sagaModel.addUnsentCommand(cmd);
        };
        var calledRemoveTimeout = false;
        var orgRemoveTimeout = sagaModel.removeTimeout;
        sagaModel.removeTimeout = function () {
          calledRemoveTimeout = true;
          orgRemoveTimeout.bind(sagaModel)();
        };
        sagaModel.commit = function (clb) {
          var cmds = calledAddCommandToSend ? sagaModel.getUndispatchedCommands() : [];

          function onAfterCommit(err) {
            if (err) return clb(err);

            function setCommandToDispatched (c, cb) {
              debug('set command to dispatched');
              self.setCommandToDispatched(dotty.get(c, self.definitions.command.id), sagaModel.id, function (err) {
                if (err) {
                  return cb(err);
                }
                sagaModel.removeUnsentCommand(c);
                cb(null);
              });
            }

            if (!self.onCommandHandle) return clb(null);

            async.each(sagaModel.getUndispatchedCommands(), function (cmd, callback) {
              debug('publish a command');
              self.onCommandHandle(cmd, function (err) {
                if (err) {
                  debug(err);
                  return callback(err);
                }
                setCommandToDispatched(cmd, callback);
              });
            }, clb);
          }

          async.each(cmds, function (cmd, fn) {
            if (dotty.exists(cmd, self.definitions.command.id)) {
              return fn(null);
            }

            self.getNewId(function (err, id) {
              if (err) {
                debug(err);
                return fn(err);
              }
              dotty.put(cmd, self.definitions.command.id, id);
              fn(null);
            });
          }, function (err) {
            if (err) {
              debug(err);
              return callback(err);
            }

            if (sagaModel.isDestroyed()) {
              self.removeSaga(sagaModel, onAfterCommit);
            } else if (calledAddCommandToSend) {
              sagaModel.setCommitStamp(new Date());
              var undispCmds = _.map(sagaModel.getUndispatchedCommands(), function (c) {
                return { id: dotty.get(c, self.definitions.command.id), payload: c };
              });
              self.sagaStore.save(sagaModel.toJSON(), undispCmds, onAfterCommit);
            } else if (calledRemoveTimeout) {
              sagaModel.setCommitStamp(new Date());
              self.sagaStore.save(sagaModel.toJSON(), [], onAfterCommit);
            } else {
              var err = new Error('Use commit only to remove a saga or to addCommandToSend!');
              debug(err);
              if (clb) { return clb(err); }
              throw err;
            }
          });
        };
        sagaModels.push(sagaModel);
      });

      callback(null, sagaModels);
    });
  },

  /**
   * Use this function to get all sagas that are older then the passed date.
   * @param {Date}     date     The date
   * @param {Function} callback The function, that will be called when this action is completed.
   *                            `function(err, sagas){}` saga is of type Array.
   */
  getOlderSagas: function (date, callback) {
    var self = this;

    this.sagaStore.getOlderSagas(date, function (err, sagas) {
      if (err) {
        debug(err);
        return callback(err);
      }
      var sagaModels = [];
      sagas.forEach(function (s) {
        var sagaModel = new SagaModel(s.id);
        sagaModel.set(s);
        sagaModel.actionOnCommit = 'update';
        var calledRemoveTimeout = false;
        var orgRemoveTimeout = sagaModel.removeTimeout;
        sagaModel.removeTimeout = function () {
          calledRemoveTimeout = true;
          orgRemoveTimeout.bind(sagaModel)();
        };
        sagaModel.commit = function (clb) {
          if (sagaModel.isDestroyed()) {
            self.removeSaga(sagaModel, clb);
          } else if (calledRemoveTimeout) {
            sagaModel.setCommitStamp(new Date());
            self.sagaStore.save(sagaModel.toJSON(), [], clb);
          } else {
            var err = new Error('Use commit only to remove a saga!');
            debug(err);
            if (clb) { return clb(err); }
            throw err;
          }
        };
        sagaModels.push(sagaModel);
      });

      callback(null, sagaModels);
    });
  },

  /**
   * Use this function to get all undispatched commands.
   * @param {Function} callback The function, that will be called when this action is completed.
   *                            `function(err, cmdsSagaMap){}` cmdsSagaMap is of type Array.
   */
  getUndispatchedCommands: function (options, callback) {
    if (!callback) {
      callback = options;
      options = {};
    }
    this.sagaStore.getUndispatchedCommands(options, callback);
  },

  /**
   * Use this function to mark a command as dispatched. (will remove it from the db)
   * @param {String}   cmdId    The command id
   * @param {String}   sagaId   The saga id
   * @param {Function} callback The function, that will be called when this action is completed. [optional]
   *                            `function(err){}`
   */
  setCommandToDispatched: function (cmdId, sagaId, callback) {
    this.sagaStore.setCommandToDispatched(cmdId, sagaId, callback);
  },

  /**
   * Use this function to remove the matched saga.
   * @param {String}   saga     The id of the saga or the saga itself
   * @param {Function} callback The function, that will be called when this action is completed. [optional]
   *                             `function(err){}`
   */
  removeSaga: function (saga, callback) {
    var sagaId = saga.id || saga;
    this.sagaStore.remove(sagaId, callback);
  },

  /**
   * Gets the last event.
   * @param {Function} callback The function, that will be called when this action is completed.
   *                            `function(err, evt){}` evt is of type Object.
   */
  getLastEvent: function (callback) {
    if (!this.revisionGuardStore)
      return callback(null, null);
    this.revisionGuardStore.getLastEvent(callback);
  }

});

module.exports = ProcessManagement;
